src/meta_schedule/task_scheduler/task_scheduler.cc

```diff
@@ -21,83 +21,225 @@
   }
-  task->search_strategy.value()->PreTuning(design_spaces, database, cost_model);
+  self->builder_results = builder->Build(inputs);
 }
 
-void TaskSchedulerNode::Tune() {
-  int n_tasks = this->tasks.size();
-  for (int task_id = 0; task_id < n_tasks; ++task_id) {
-    InitializeTask(task_id);
+void SendToRunner(TaskRecordNode* self, const Runner& runner) {
+  auto _ = Profiler::TimedScope("SendToRunner");
+  Array<MeasureCandidate> candidates = self->measure_candidates.value();
+  Array<BuilderResult> builder_results = self->builder_results.value();
+  Target target = self->ctx->target.value();
+  ICHECK_EQ(candidates.size(), builder_results.size());
+  int n = candidates.size();
+  int n_build_errors = 0;
+  Array<RunnerInput> inputs;
+  inputs.reserve(n);
+  for (int i = 0; i < n; ++i) {
+    const MeasureCandidate& candidate = candidates[i];
+    const BuilderResult& builder_result = builder_results[i];
+    if (builder_result->error_msg.defined()) {
+      ++n_build_errors;
+      continue;
+    }
+    inputs.push_back(RunnerInput(/*artifact_path=*/builder_result->artifact_path.value(),
+                                 /*device_type=*/target->kind->name,
+                                 /*args_info=*/candidate->args_info));
+  }
+  Array<RunnerFuture> futures = runner->Run(inputs);
+  if (n_build_errors == 0) {
+    self->runner_futures = futures;
+    return;
+  }
+  Array<RunnerFuture> results;
+  results.reserve(n);
+  for (int i = 0, j = 0; i < n; ++i) {
+    const BuilderResult& builder_result = builder_results[i];
+    if (builder_result->error_msg.defined()) {
+      results.push_back(RunnerFuture(
+          /*f_done=*/[]() -> bool { return true; },
+          /*f_result=*/
+          [msg = builder_result->error_msg]() -> RunnerResult {
+            return RunnerResult(NullOpt, msg);
+          }));
+    } else {
+      results.push_back(futures[j++]);
+    }
+  }
+  self->runner_futures = results;
+}
+
+void TaskCleanUp(TaskRecordNode* self, int task_id, const Array<RunnerResult>& results) {
+  ICHECK_EQ(self->builder_results.value().size(), results.size());
+  ICHECK_EQ(self->runner_futures.value().size(), results.size());
+  int n = results.size();
+  std::string name = self->ctx->task_name.value();
+  const PackedFunc& logger = self->ctx->logger;
+  for (int i = 0; i < n; ++i) {
+    const BuilderResult& builder_result = self->builder_results.value()[i];
+    const MeasureCandidate& candidate = self->measure_candidates.value()[i];
+    const RunnerResult& runner_result = results[i];
+    Optional<String> error_msg = NullOpt;
+    int trials = self->latency_ms.size() + 1;
+    double run_ms = 1e9;
+    if ((error_msg = builder_result->error_msg)) {
+      ++self->build_error_count;
+    } else if ((error_msg = runner_result->error_msg)) {
+      ++self->run_error_count;
+    } else {
+      run_ms = GetRunMsMedian(runner_result);
+    }
+    self->latency_ms.push_back(run_ms);
+    if (error_msg) {
```

Comment:
I'm wondering how frequent these errors happen during tuning. Should we print all errors or just errors above count threshold?

Reply:
It could be pretty frequent if something goes wrong, but normally it didn't happen often. By default, the error message is only sent to a per-task logging file rather than the main screen, so I wouldn't worry too much about it